File: server/subtitles.py
from __future__ import annotations

from typing import Iterable, Literal, Protocol


SubtitleFormat = Literal["srt", "vtt"]
SubtitleTrack = Literal["source", "target", "bilingual"]


class SubtitleSegment(Protocol):
    id: str
    start_ms: int
    end_ms: int
    source_text: str
    target_text: str


def _timestamp(milliseconds: int, separator: str) -> str:
    hours, remainder = divmod(max(0, milliseconds), 3_600_000)
    minutes, remainder = divmod(remainder, 60_000)
    seconds, millis = divmod(remainder, 1_000)
    return f"{hours:02d}:{minutes:02d}:{seconds:02d}{separator}{millis:03d}"


def _clean(value: str) -> str:
    return value.replace("\r\n", "\n").replace("\r", "\n").strip()


def _cue_text(segment: SubtitleSegment, track: SubtitleTrack) -> str:
    source = _clean(segment.source_text)
    target = _clean(segment.target_text)
    if track == "source":
        return source
    if track == "bilingual":
        return f"{source}\n{target}"
    return target


def render_subtitles(
    segments: Iterable[SubtitleSegment],
    subtitle_format: SubtitleFormat,
    track: SubtitleTrack,
) -> str:
    cues: list[str] = []
    for segment in segments:
        if segment.end_ms <= segment.start_ms:
            raise ValueError(f"invalid_segment_timing:{segment.id}")
        text = _cue_text(segment, track)
        if not text:
            continue
        cue_number = len(cues) + 1
        if subtitle_format == "srt":
            timing = (
                f"{_timestamp(segment.start_ms, ',')} --> "
                f"{_timestamp(segment.end_ms, ',')}"
            )
            cues.append(f"{cue_number}\n{timing}\n{text}")
        else:
            timing = (
                f"{_timestamp(segment.start_ms, '.')} --> "
                f"{_timestamp(segment.end_ms, '.')}"
            )
            cues.append(f"{segment.id}\n{timing}\n{text}")

    body = "\n\n".join(cues) + ("\n" if cues else "")
    if subtitle_format == "vtt":
        return f"WEBVTT\n\n{body}"
    # BOM improves UTF-8 Chinese detection in common Windows subtitle players.
    return f"\ufeff{body}"

File: src/subtitles.ts
import type { Segment } from "./types";

export type SubtitleFormat = "srt" | "vtt";
export type SubtitleTrack = "target" | "bilingual" | "source";

function parseClock(value: string) {
  const parts = value.trim().split(":").map(Number);
  if (parts.some((part) => !Number.isFinite(part))) {
    throw new Error(`invalid_subtitle_timestamp:${value}`);
  }
  const seconds = parts.pop() ?? 0;
  const minutes = parts.pop() ?? 0;
  const hours = parts.pop() ?? 0;
  return Math.round((hours * 3600 + minutes * 60 + seconds) * 1000);
}

function timestamp(milliseconds: number, separator: "," | ".") {
  const safe = Math.max(0, milliseconds);
  const hours = Math.floor(safe / 3_600_000);
  const minutes = Math.floor((safe % 3_600_000) / 60_000);
  const seconds = Math.floor((safe % 60_000) / 1_000);
  const millis = safe % 1_000;
  return `${String(hours).padStart(2, "0")}:${String(minutes).padStart(2, "0")}:${String(seconds).padStart(2, "0")}${separator}${String(millis).padStart(3, "0")}`;
}

function cueText(segment: Segment, track: SubtitleTrack) {
  const source = segment.source.replaceAll("\r", "").trim();
  const target = segment.target.replaceAll("\r", "").trim();
  if (track === "source") return source;
  if (track === "bilingual") return `${source}\n${target}`;
  return target;
}

export function serializeSubtitles(
  segments: Segment[],
  format: SubtitleFormat,
  track: SubtitleTrack,
) {
  const cues: string[] = [];
  for (const segment of segments) {
    const start = parseClock(segment.start);
    const end = parseClock(segment.end);
    if (end <= start) throw new Error(`invalid_segment_timing:${segment.id}`);
    const text = cueText(segment, track);
    if (!text) continue;
    const separator = format === "srt" ? "," : ".";
    const timing = `${timestamp(start, separator)} --> ${timestamp(end, separator)}`;
    cues.push(
      format === "srt"
        ? `${cues.length + 1}\n${timing}\n${text}`
        : `${segment.id.replaceAll(" ", "")}\n${timing}\n${text}`,
    );
  }
  const body = `${cues.join("\n\n")}${cues.length ? "\n" : ""}`;
  return format === "vtt" ? `WEBVTT\n\n${body}` : `\ufeff${body}`;
}

export function subtitleFilename(
  sourceFilename: string,
  format: SubtitleFormat,
  track: SubtitleTrack,
) {
  const base = sourceFilename.replace(/\.[^.]+$/, "").replace(/[<>:"/\\|?*]/g, "_");
  const language = track === "source" ? "zh" : track === "target" ? "en" : "zh-en";
  return `${base}.${language}.${format}`;
}
